Add git repo validation and build tool detection using git-service

rohitkrai03 opened the pull request:
Related Story - https://jira.coreos.com/browse/ODC-1155

This PR - 
- Adds validation of git repo using git service.
- Adds detection and recommendation of builder images based on the git repo using git service.

Gif - 
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/6041994/66786969-c58e0180-eeff-11e9-93e2-073d6f1369f3.gif)

rohitkrai03 commented:
/test e2e-aws-console-olm

rohitkrai03 commented:
/test e2e-aws-console-olm

rohitkrai03 commented:
> For the first-time builder image recommends by applying the border and star icon on the corner of the card. But when I change the git URL then star icon move to another detected builder image but border remains to the previous detected builder image.
> 
> ![Peek 2019-10-15 12-15](https://user-images.githubusercontent.com/2561818/66807314-846d1000-ef46-11e9-96a6-91bd4b470819.gif)

This is desired behaviour since we do not want to select a builder image if user has already selected one. In this case it was pre-selected by recommendation earlier.

vikram-raj commented:
> This is desired behaviour since we do not want to select a builder image if user has already selected one. In this case it was pre-selected by recommendation earlier.

@rohitkrai03 I am selecting the builder image. For the first time border get applied automatically.
Is the border applied only when the user selects the builder image?

rohitkrai03 commented:
lo

> > This is desired behaviour since we do not want to select a builder image if user has already selected one. In this case it was pre-selected by recommendation earlier.
> 
> @rohitkrai03 I am selecting the builder image. For the first time border get applied automatically.
> Is the border applied only when the user selects the builder image?

If user has not selected any builder image then while recommending we also select the recommended builder image, this is why you see blue border at first since you have not selected anything. But the second time you already have a selected builder image that was selected by recommendation, since we cannot distinguish if it was a user selection or recommendation selection for the second time, the builder image is not selected again. It retains the old value.

rohitkrai03 commented:
> Getting this error
> `Uncaught (in promise) TypeError: Cannot read property 'isRepoReachable' of null`
> 
> When I leave the Git URL field without entering any value.

Fixed

vikram-raj commented:
@rohitkrai03 No, as per this PR we are not selecting the builder image until the user selects one. Because I am not seeing the builder image version dropdown.

rohitkrai03 commented:
> @rohitkrai03 No, as per this PR we are not selecting the builder image until the user selects one. Because I am not seeing the builder image version dropdown.

Fixed. I missed the tag selection for builder image earlier.

debsmita1 commented:
At first I entered `https://github.com/rails/rails` as the git url and on validation it detects `Ruby` builder image. On updating the url I saw this :
![validating](https://user-images.githubusercontent.com/22490998/66859952-ddba5b00-efa9-11e9-84fc-02701add3c28.png)

I feel we shouldn't show the `Builder Image detected` notification when the url is being validated. What do you think ?

debsmita1 commented:
Do you think we should hide the notification if user selects a builder image after being notified ?

![user selection](https://user-images.githubusercontent.com/22490998/66860151-3a1d7a80-efaa-11e9-891c-49241d3ee8aa.png)

debsmita1 commented on frontend/packages/dev-console/src/components/import/git/GitSection.tsx:
```diff
@@ -31,13 +50,44 @@ const GitSection: React.FC<GitSectionProps> = ({ showSample }) => {
+      setFieldValue('image.recommended', '');
+      setFieldValue('image.couldNotRecommend', false);
+      setFieldValue('git.isUrlValidated', false);
+      setFieldError('git.url', 'Git respository is not reachable');
```
```suggestion
      setFieldError('git.url', 'Git respository is not reachable.');

rohitkrai03 replied:
Done

rohitkrai03 commented:
@debsmita1 Both of your comments related to alert notifications should be decided by UX and since I couldn't find anything related to this in the design doc i think its okay for now. If UX wants us to change this behaviour we can add it later as a fix.

christianvogt commented:
I think we should start clearing out the validation and recommended states when the user goes back and edits the git url. On change before they even blur the field.

christianvogt commented:
Validation message doesn't match design:
![image](https://user-images.githubusercontent.com/14068621/66885287-dbec9980-efa1-11e9-8469-2fa988de60e8.png)
![image](https://user-images.githubusercontent.com/14068621/66885291-e0b14d80-efa1-11e9-9b5f-609cd648e197.png)

Recommended star color is very dark compared to designs:
![image](https://user-images.githubusercontent.com/14068621/66885317-f3c41d80-efa1-11e9-87fd-4c5fa28e3f8e.png)
![image](https://user-images.githubusercontent.com/14068621/66885322-f9b9fe80-efa1-11e9-94d3-0959f0c9f411.png)

rohitkrai03 commented:
> I think we should start clearing out the validation and recommended states when the user goes back and edits the git url. On change before they even blur the field.

Done.

rohitkrai03 commented:
> Validation message doesn't match design:
> ![image](https://user-images.githubusercontent.com/14068621/66885287-dbec9980-efa1-11e9-8469-2fa988de60e8.png)
> ![image](https://user-images.githubusercontent.com/14068621/66885291-e0b14d80-efa1-11e9-9b5f-609cd648e197.png)
> 
> Recommended star color is very dark compared to designs:
> ![image](https://user-images.githubusercontent.com/14068621/66885317-f3c41d80-efa1-11e9-87fd-4c5fa28e3f8e.png)
> ![image](https://user-images.githubusercontent.com/14068621/66885322-f9b9fe80-efa1-11e9-94d3-0959f0c9f411.png)

Updated the color of star icon from `--pf-global--primary-color--200` to `--pf-global--primary-color--100`. For the validation texts below git url input field I've added the check icon temporarily until https://github.com/patternfly/patternfly-next/issues/2320 is implemented. I've made the text bolder for now.

rohitkrai03 commented:
/retest

rohitkrai03 commented:
/test e2e-aws-console
